test/new-e2e/tests/windows/install-test/install_test.go

```diff
@@ -110,40 +116,136 @@ func (is *agentMSISuite) TestUpgrade() {
 	is.prepareHost()
 
-	t, err := NewTester(is.T(), vm, WithExpectedAgentVersion(is.agentPackage.AgentVersion()))
+	_ = is.installLastStable(vm, "", filepath.Join(outputDir, "install.log"))
+
+	t, err := NewTester(is.T(), vm,
+		WithAgentPackage(is.agentPackage),
+	)
 	is.Require().NoError(err, "should create tester")
 
-	// install old agent
-	_ = is.installLastStable(t, filepath.Join(outputDir, "install.log"))
+	if !is.Run(fmt.Sprintf("upgrade to %s", t.agentPackage.AgentVersion()), func() {
+		err = t.InstallAgent(is.T(), "", filepath.Join(outputDir, "upgrade.log"))
+		is.Require().NoError(err, "should upgrade to agent %s", t.agentPackage.AgentVersion())
+	}) {
+		is.T().FailNow()
+	}
 
-	// upgrade to new agent
-	if !t.TestInstallAgentPackage(is.T(), is.agentPackage, "", filepath.Join(outputDir, "upgrade.log")) {
-		is.T().Fatal("failed to upgrade agent")
+	if !t.TestExpectations(is.T()) {
+		is.T().FailNow()
 	}
 
-	t.TestRuntimeExpectations(is.T())
 	t.TestUninstall(is.T(), filepath.Join(outputDir, "uninstall.log"))
 }
 
-// This is separate from TestInstallAgentPackage because previous versions of the agent
-// may not conform to the latest test expectations.
-func (is *agentMSISuite) installLastStable(t *Tester, logfile string) *windowsAgent.Package {
-	var agentPackage *windowsAgent.Package
+// TC-INS-002
+func (is *agentMSISuite) TestUpgradeRollback() {
+	outputDir, err := runner.GetTestOutputDir(runner.GetProfile(), is.T())
+	is.Require().NoError(err, "should get output dir")
+	is.T().Logf("Output dir: %s", outputDir)
 
-	if !is.Run("install prev stable agent", func() {
-		var err error
+	vm := is.Env().RemoteHost
+	is.prepareHost()
 
-		agentPackage, err = windowsAgent.GetLastStablePackageFromEnv()
-		is.Require().NoError(err, "should get last stable agent package from env")
+	previousTester := is.installLastStable(vm, "", filepath.Join(outputDir, "install.log"))
 
-		t.InstallAgentPackage(is.T(), agentPackage, "", logfile)
+	t, err := NewTester(is.T(), vm,
+		WithAgentPackage(is.agentPackage),
+	)
+	is.Require().NoError(err, "should create tester")
 
-		agentVersion, err := t.InstallTestClient.GetAgentVersion()
-		is.Require().NoError(err, "should get agent version")
-		windowsAgent.TestAgentVersion(is.T(), agentPackage.AgentVersion(), agentVersion)
+	if !is.Run(fmt.Sprintf("upgrade to %s with rollback", t.agentPackage.AgentVersion()), func() {
+		err = t.InstallAgent(is.T(), "WIXFAILWHENDEFERRED=1", filepath.Join(outputDir, "upgrade.log"))
+		is.Require().Error(err, "should fail to install agent %s", t.agentPackage.AgentVersion())
 	}) {
-		is.T().Fatal("failed to install last stable agent")
+		is.T().FailNow()
+	}
+
+	// TODO: we shouldn't have to start the agent manually after rollback
+	//       but the kitchen tests did too.
+	err = windows.StartService(t.host, "DatadogAgent")
+	is.Require().NoError(err, "agent service should start after rollback")
+
+	if !previousTester.TestExpectations(is.T()) {
+		is.T().FailNow()
```

Comment:
I've been going back and forth on this actually.

I like that with the `FailNow` structure it easy to see that the function's control flow may change, and it has a similar structure to the standard `if err != nil { return err } ` pattern.

I like `require` because it's shorter, but `require.True(TestExpectations())` or `require.True(InstallAgent())` feel funny and don't seem to convey the same sort of meaning that you would get when using `require.Error()` or `require.Equal()`.